core/src/main/java/com/linecorp/armeria/server/http/auth/HttpAuthService.java

```diff
@@ -91,10 +92,20 @@ protected HttpResponse onFailure(ServiceRequestContext ctx, HttpRequest req) thr
 
     @Override
     public HttpResponse serve(ServiceRequestContext ctx, HttpRequest req) throws Exception {
-        if (authorize(req.headers())) {
-            return onSuccess(ctx, req);
-        } else {
-            return onFailure(ctx, req);
-        }
+        DeferredHttpResponse res = new DeferredHttpResponse();
+
+        authorize(req, ctx).whenComplete((result, t) -> {
+            try {
+                if (t != null || !result) {
+                    res.delegate(onFailure(ctx, req));
```

Comment:
Seems like `t` goes nowhere. How about adding the third nullable parameter to `onFailure`?